File: frontend/src/services/admissionService.js
import apiClient from './apiClient';
import { API_ENDPOINTS } from '../utils/constants';
import {
  IS_PREVIEW_MODE,
  MOCK_ADMISSION_APPLICATIONS,
  MOCK_ADMISSION_DOCUMENTS,
} from '../utils/mockData';

// Local mutable state for mock session interactions
let localApplications = [...MOCK_ADMISSION_APPLICATIONS];
let localDocuments = [...MOCK_ADMISSION_DOCUMENTS];

export const admissionService = {
  /**
   * Get institutional admissions dashboard statistics
   * GET /api/v1/admission/dashboard
   */
  getDashboard: async () => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.ADMISSION.DASHBOARD);
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const total = localApplications.length;
        const submitted = localApplications.filter((a) => a.status === 'SUBMITTED').length;
        const underReview = localApplications.filter((a) => a.status === 'UNDER_REVIEW').length;
        const approved = localApplications.filter((a) => a.status === 'APPROVED').length;
        const rejected = localApplications.filter((a) => a.status === 'REJECTED').length;
        return {
          totalApplications: total,
          submitted,
          underReview,
          approved,
          rejected,
        };
      }
      throw err;
    }
  },

  /**
   * List paginated admission applications with optional status filter
   * GET /api/v1/admission/applications
   */
  getApplications: async ({ page = 0, size = 50, status = null } = {}) => {
    try {
      const params = { page, size };
      if (status && status !== 'ALL') {
        params.status = status;
      }
      const response = await apiClient.get(API_ENDPOINTS.ADMISSION.APPLICATIONS, { params });
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        let filtered = [...localApplications];
        if (status && status !== 'ALL') {
          filtered = filtered.filter((a) => a.status === status);
        }
        return {
          content: filtered,
          totalElements: filtered.length,
          totalPages: 1,
          size,
          number: page,
        };
      }
      throw err;
    }
  },

  /**
   * Get a single admission application by UUID
   * GET /api/v1/admission/applications/{id}
   */
  getApplicationById: async (id) => {
    try {
      const response = await apiClient.get(`${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${id}`);
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const found = localApplications.find((a) => a.id === id);
        if (!found) throw new Error('Application not found');
        return found;
      }
      throw err;
    }
  },

  /**
   * Move application to UNDER_REVIEW
   * PATCH /api/v1/admission/applications/{id}/review
   */
  reviewApplication: async (id) => {
    try {
      const response = await apiClient.patch(`${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${id}/review`);
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const index = localApplications.findIndex((a) => a.id === id);
        if (index === -1) throw new Error('Application not found');
        const updated = {
          ...localApplications[index],
          status: 'UNDER_REVIEW',
          reviewedAt: new Date().toISOString(),
          reviewedBy: 'ADM-OFFICER',
        };
        localApplications[index] = updated;
        return updated;
      }
      throw err;
    }
  },

  /**
   * Approve application
   * PATCH /api/v1/admission/applications/{id}/approve
   */
  approveApplication: async (id) => {
    try {
      const response = await apiClient.patch(`${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${id}/approve`);
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const index = localApplications.findIndex((a) => a.id === id);
        if (index === -1) throw new Error('Application not found');
        const updated = {
          ...localApplications[index],
          status: 'APPROVED',
          reviewedAt: new Date().toISOString(),
          reviewedBy: 'ADM-OFFICER',
          rejectionReason: null,
        };
        localApplications[index] = updated;
        return updated;
      }
      throw err;
    }
  },

  /**
   * Reject application with reason
   * PATCH /api/v1/admission/applications/{id}/reject
   */
  rejectApplication: async (id, reason) => {
    try {
      const response = await apiClient.patch(
        `${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${id}/reject`,
        { reason }
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const index = localApplications.findIndex((a) => a.id === id);
        if (index === -1) throw new Error('Application not found');
        const updated = {
          ...localApplications[index],
          status: 'REJECTED',
          rejectionReason: reason,
          reviewedAt: new Date().toISOString(),
          reviewedBy: 'ADM-OFFICER',
        };
        localApplications[index] = updated;
        return updated;
      }
      throw err;
    }
  },

  /**
   * Get documents for an application
   * GET /api/v1/admission/applications/{id}/documents
   */
  getDocuments: async (applicationId) => {
    try {
      const response = await apiClient.get(
        `${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${applicationId}/documents`
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        return localDocuments.filter((d) => d.applicationId === applicationId);
      }
      throw err;
    }
  },

  /**
   * Add a document to an application
   * POST /api/v1/admission/applications/{id}/documents
   */
  addDocument: async (applicationId, docData) => {
    try {
      const response = await apiClient.post(
        `${API_ENDPOINTS.ADMISSION.APPLICATIONS}/${applicationId}/documents`,
        docData
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const newDoc = {
          id: `doc-${Date.now()}-${Math.floor(Math.random() * 1000)}`,
          applicationId,
          documentType: docData.documentType,
          fileName: docData.fileName || 'Uploaded_Document.pdf',
          fileUrl: docData.fileUrl || 'https://example.com/docs/file.pdf',
          uploadedAt: new Date().toISOString(),
          verified: docData.verified || false,
        };
        localDocuments.push(newDoc);
        return newDoc;
      }
      throw err;
    }
  },

  /**
   * Verify or unverify a candidate document
   * PATCH /api/v1/admission/documents/{id}/verify
   */
  verifyDocument: async (documentId, verified = true) => {
    try {
      const response = await apiClient.patch(
        `${API_ENDPOINTS.ADMISSION.BASE}/documents/${documentId}/verify`,
        null,
        { params: { verified } }
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const index = localDocuments.findIndex((d) => d.id === documentId);
        if (index !== -1) {
          localDocuments[index] = { ...localDocuments[index], verified };
          return localDocuments[index];
        }
        return { id: documentId, verified };
      }
      throw err;
    }
  },

  /**
   * Publicly track application by application number or email
   * GET /api/v1/admission/public/track/{query}
   */
  trackApplication: async (query) => {
    try {
      const trimmed = (query || '').trim();
      const response = await apiClient.get(
        `${API_ENDPOINTS.ADMISSION.PUBLIC_TRACK}/${encodeURIComponent(trimmed)}`
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const q = (query || '').trim().toLowerCase();
        const found = localApplications.find(
          (a) =>
            a.applicationNumber?.toLowerCase() === q ||
            a.id?.toLowerCase() === q ||
            a.email?.toLowerCase() === q
        );
        if (!found) {
          throw new Error(
            `No admission application found matching "${query}". Please check your application reference or email.`
          );
        }
        return found;
      }
      throw err;
    }
  },

  /**
   * Publicly get documents for tracked application
   * GET /api/v1/admission/public/track/{query}/documents
   */
  trackDocuments: async (query) => {
    try {
      const trimmed = (query || '').trim();
      const response = await apiClient.get(
        `${API_ENDPOINTS.ADMISSION.PUBLIC_TRACK}/${encodeURIComponent(trimmed)}/documents`
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const q = (query || '').trim().toLowerCase();
        const found = localApplications.find(
          (a) =>
            a.applicationNumber?.toLowerCase() === q ||
            a.id?.toLowerCase() === q ||
            a.email?.toLowerCase() === q
        );
        if (found) {
          return localDocuments.filter((d) => d.applicationId === found.id);
        }
        return [];
      }
      throw err;
    }
  },

  /**
   * Submit public application by college code
   * POST /api/v1/admission/public/code/{collegeCode}/apply
   */
  applyByCode: async (collegeCode, applicationData) => {
    try {
      const response = await apiClient.post(
        `${API_ENDPOINTS.ADMISSION.PUBLIC_APPLY_CODE}/${collegeCode}/apply`,
        applicationData
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        const newApp = {
          id: `adm-app-${Date.now()}`,
          applicationNumber: `ADM-2026-${Math.floor(10000 + Math.random() * 90000)}`,
          collegeCode: collegeCode || 'COLLEGE',
          collegeId: applicationData.collegeId || 'COL-DELHI-001',
          applicantName: applicationData.applicantName,
          email: applicationData.email,
          phoneNumber: applicationData.phoneNumber,
          gender: applicationData.gender || 'NOT_SPECIFIED',
          dateOfBirth: applicationData.dateOfBirth,
          fatherName: applicationData.fatherName,
          motherName: applicationData.motherName,
          address: applicationData.address,
          courseId: applicationData.courseId,
          courseName: applicationData.courseName,
          branchId: applicationData.branchId,
          branchName: applicationData.branchName,
          previousQualification: applicationData.previousQualification,
          previousInstitution: applicationData.previousInstitution,
          previousPercentage: Number(applicationData.previousPercentage) || 0,
          status: 'SUBMITTED',
          submittedAt: new Date().toISOString(),
          reviewedAt: null,
          reviewedBy: null,
          rejectionReason: null,
        };
        localApplications.unshift(newApp);
        return newApp;
      }
      throw err;
    }
  },

  /**
   * Submit public application by college ID
   * POST /api/v1/admission/public/id/{collegeId}/apply
   */
  applyById: async (collegeId, applicationData) => {
    try {
      const response = await apiClient.post(
        `${API_ENDPOINTS.ADMISSION.PUBLIC_APPLY_ID}/${collegeId}/apply`,
        applicationData
      );
      return response.data;
    } catch (err) {
      if (IS_PREVIEW_MODE) {
        return admissionService.applyByCode('COLLEGE', { ...applicationData, collegeId });
      }
      throw err;
    }
  },
};

File: backend/AdmissionService/AdmissionService/src/main/java/com/example/admission/admission/service/AdmissionService.java
package com.example.admission.admission.service;

import com.example.admission.admission.entity.*;
import com.example.admission.admission.repository.*;
import com.example.admission.admission.request.*;
import com.example.admission.admission.response.*;
import com.example.admission.common.context.*;
import com.example.admission.core.client.CoreCollegeClient;
import com.example.admission.core.dto.CollegePublicResponse;
import com.example.admission.notification.NotificationProducer;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.*;
import java.util.*;

@Service
@RequiredArgsConstructor
public class AdmissionService {
    private final AdmissionApplicationRepository appRepo;
    private final AdmissionDocumentRepository docRepo;
    private final CoreCollegeClient core;
    private final NotificationProducer notificationProducer;

    @Transactional
    public AdmissionApplicationResponse applyByCode(String code, AdmissionApplicationRequest r) {
        CollegePublicResponse c = core.getCollegeByCode(code);
        if (c == null || c.getCollegeId() == null) throw new IllegalArgumentException("College not found");
        AdmissionApplication a = build(c.getCollegeId(), c.getCollegeCode(), r);
        AdmissionApplication saved = appRepo.save(a);
        notificationProducer.send(
                saved.getEmail(),
                "Admission Application Submitted - " + saved.getApplicationNumber(),
                "Dear " + saved.getApplicantName() + ",\n\nYour admission application " + saved.getApplicationNumber() + " has been submitted successfully to " + c.getCollegeName() + ".\n\nCurrent status: SUBMITTED."
        );
        return map(saved);
    }

    @Transactional
    public AdmissionApplicationResponse applyById(UUID id, AdmissionApplicationRequest r) {
        CollegePublicResponse c = core.getCollegeById(id);
        if (c == null || c.getCollegeId() == null) throw new IllegalArgumentException("College not found");
        AdmissionApplication a = build(c.getCollegeId(), c.getCollegeCode(), r);
        AdmissionApplication saved = appRepo.save(a);
        notificationProducer.send(
                saved.getEmail(),
                "Admission Application Submitted - " + saved.getApplicationNumber(),
                "Dear " + saved.getApplicantName() + ",\n\nYour admission application " + saved.getApplicationNumber() + " has been submitted successfully to " + c.getCollegeName() + ".\n\nCurrent status: SUBMITTED."
        );
        return map(saved);
    }

    private AdmissionApplication build(UUID collegeId, String code, AdmissionApplicationRequest r) {
        return AdmissionApplication.builder().applicationNumber("ADM-" + Year.now().getValue() + "-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase()).collegeId(collegeId).collegeCode(code).applicantName(r.getApplicantName()).email(r.getEmail()).phoneNumber(r.getPhoneNumber()).dateOfBirth(r.getDateOfBirth()).gender(r.getGender()).address(r.getAddress()).fatherName(r.getFatherName()).motherName(r.getMotherName()).courseId(r.getCourseId()).courseName(r.getCourseName()).branchId(r.getBranchId()).branchName(r.getBranchName()).previousQualification(r.getPreviousQualification()).previousInstitution(r.getPreviousInstitution()).previousPercentage(r.getPreviousPercentage()).status(AdmissionStatus.SUBMITTED).submittedAt(LocalDateTime.now()).build();
    }

    public Page<AdmissionApplicationResponse> list(int page, int size, AdmissionStatus status) {
        UUID cid = requireCollege();
        Pageable p = PageRequest.of(Math.max(page, 0), Math.min(Math.max(size, 1), 100));
        Page<AdmissionApplication> a = status == null ? appRepo.findByCollegeIdOrderBySubmittedAtDesc(cid, p) : appRepo.findByCollegeIdAndStatusOrderBySubmittedAtDesc(cid, status, p);
        return a.map(this::map);
    }

    public AdmissionApplicationResponse get(UUID id) {
        return map(tenant(id));
    }

    @Transactional
    public AdmissionApplicationResponse review(UUID id) {
        AdmissionApplication a = tenant(id);
        if (a.getStatus() != AdmissionStatus.SUBMITTED)
            throw new IllegalStateException("Only submitted applications can be moved to review");
        a.setStatus(AdmissionStatus.UNDER_REVIEW);
        a.setReviewedBy(UserContext.getUserId());
        a.setReviewedAt(LocalDateTime.now());
        return map(appRepo.save(a));
    }

    @Transactional
    public AdmissionApplicationResponse approve(UUID id) {
        AdmissionApplication a = tenant(id);
        ensureOpen(a);
        a.setStatus(AdmissionStatus.APPROVED);
        a.setReviewedBy(UserContext.getUserId());
        a.setReviewedAt(LocalDateTime.now());
        AdmissionApplication saved = appRepo.save(a);
        notificationProducer.send(
                saved.getEmail(),
                "Admission Application Approved - " + saved.getApplicationNumber(),
                "Dear " + saved.getApplicantName() + ",\n\nCongratulations. Your admission application " + saved.getApplicationNumber() + " has been approved."
        );
        return map(saved);
    }

    @Transactional
    public AdmissionApplicationResponse reject(UUID id, RejectAdmissionRequest req) {
        AdmissionApplication a = tenant(id);
        ensureOpen(a);
        a.setStatus(AdmissionStatus.REJECTED);
        a.setRejectionReason(req.getReason());
        a.setReviewedBy(UserContext.getUserId());
        a.setReviewedAt(LocalDateTime.now());
        AdmissionApplication saved = appRepo.save(a);
        notificationProducer.send(
                saved.getEmail(),
                "Admission Application Rejected - " + saved.getApplicationNumber(),
                "Dear " + saved.getApplicantName() + ",\n\nYour admission application " + saved.getApplicationNumber() + " has been rejected.\n\nReason: " + (saved.getRejectionReason() == null ? "Not specified" : saved.getRejectionReason())
        );
        return map(saved);
    }

    public DashboardResponse dashboard() {
        UUID c = requireCollege();
        return DashboardResponse.builder().totalApplications(appRepo.countByCollegeId(c)).submitted(appRepo.countByCollegeIdAndStatus(c, AdmissionStatus.SUBMITTED)).underReview(appRepo.countByCollegeIdAndStatus(c, AdmissionStatus.UNDER_REVIEW)).approved(appRepo.countByCollegeIdAndStatus(c, AdmissionStatus.APPROVED)).rejected(appRepo.countByCollegeIdAndStatus(c, AdmissionStatus.REJECTED)).build();
    }

    @Transactional
    public DocumentResponse addDocument(UUID id, DocumentRequest r) {
        tenant(id);
        AdmissionDocument d = AdmissionDocument.builder().applicationId(id).documentType(r.getDocumentType()).fileName(r.getFileName()).fileUrl(r.getFileUrl()).uploadedAt(LocalDateTime.now()).verified(false).build();
        return doc(docRepo.save(d));
    }

    @Transactional
    public DocumentResponse verifyDocument(UUID documentId, boolean verified) {
        AdmissionDocument d = docRepo.findById(documentId)
                .orElseThrow(() -> new IllegalArgumentException("Document not found"));
        // Verify tenant
        tenant(d.getApplicationId());
        d.setVerified(verified);
        return doc(docRepo.save(d));
    }

    public List<DocumentResponse> documents(UUID id) {
        tenant(id);
        return docRepo.findByApplicationIdOrderByUploadedAtAsc(id).stream().map(this::doc).toList();
    }

    public AdmissionApplicationResponse track(String ref) {
        AdmissionApplication a = appRepo.findByApplicationNumber(ref)
                .or(() -> appRepo.findFirstByEmailIgnoreCaseOrderBySubmittedAtDesc(ref))
                .orElseThrow(() -> new IllegalArgumentException("Application not found for reference: " + ref));
        return map(a);
    }

    public List<DocumentResponse> trackDocuments(String ref) {
        AdmissionApplication a = appRepo.findByApplicationNumber(ref)
                .or(() -> appRepo.findFirstByEmailIgnoreCaseOrderBySubmittedAtDesc(ref))
                .orElseThrow(() -> new IllegalArgumentException("Application not found for reference: " + ref));
        return docRepo.findByApplicationIdOrderByUploadedAtAsc(a.getId()).stream().map(this::doc).toList();
    }

    private AdmissionApplication tenant(UUID id) {
        AdmissionApplication a = appRepo.findById(id).orElseThrow(() -> new IllegalArgumentException("Application not found"));
        if (!requireCollege().equals(a.getCollegeId())) throw new IllegalArgumentException("Application not found");
        return a;
    }

    private UUID requireCollege() {
        UUID id = CollegeContext.getCollegeId();
        if (id == null) throw new IllegalStateException("CollegeId header is required");
        return id;
    }

    private void ensureOpen(AdmissionApplication a) {
        if (a.getStatus() == AdmissionStatus.APPROVED || a.getStatus() == AdmissionStatus.REJECTED)
            throw new IllegalStateException("Application has already been finalized");
    }

    private AdmissionApplicationResponse map(AdmissionApplication a) {
        return AdmissionApplicationResponse.builder().id(a.getId()).applicationNumber(a.getApplicationNumber()).collegeId(a.getCollegeId()).collegeCode(a.getCollegeCode()).applicantName(a.getApplicantName()).email(a.getEmail()).phoneNumber(a.getPhoneNumber()).dateOfBirth(a.getDateOfBirth()).gender(a.getGender()).address(a.getAddress()).fatherName(a.getFatherName()).motherName(a.getMotherName()).courseId(a.getCourseId()).courseName(a.getCourseName()).branchId(a.getBranchId()).branchName(a.getBranchName()).previousQualification(a.getPreviousQualification()).previousInstitution(a.getPreviousInstitution()).previousPercentage(a.getPreviousPercentage()).status(a.getStatus()).submittedAt(a.getSubmittedAt()).reviewedAt(a.getReviewedAt()).reviewedBy(a.getReviewedBy()).rejectionReason(a.getRejectionReason()).build();
    }

    private DocumentResponse doc(AdmissionDocument d) {
        return DocumentResponse.builder().id(d.getId()).applicationId(d.getApplicationId()).documentType(d.getDocumentType()).fileName(d.getFileName()).fileUrl(d.getFileUrl()).uploadedAt(d.getUploadedAt()).verified(d.getVerified()).build();
    }
}
